internal/datacoord/meta/segment/segment_info.go

```diff
@@ -0,0 +1,194 @@
+	cloned := &SegmentInfo{
+		SegmentInfo:   info,
+		currRows:      s.currRows,
+		allocations:   s.allocations,
+		lastFlushTime: s.lastFlushTime,
+		isCompacting:  s.isCompacting,
+		// cannot copy size, since binlog may be changed
+		lastWrittenTime: s.lastWrittenTime,
+	}
+	for _, opt := range opts {
+		opt.UpdateSegment(cloned)
+	}
+	return cloned
+}
+
+// ShadowClone shadow clone the segment and return a new instance
+func (s *SegmentInfo) ShadowClone(opts ...SegmentOperator) *SegmentInfo {
+	cloned := &SegmentInfo{
+		SegmentInfo:     s.SegmentInfo,
+		currRows:        s.currRows,
+		allocations:     s.allocations,
+		lastFlushTime:   s.lastFlushTime,
+		isCompacting:    s.isCompacting,
+		lastWrittenTime: s.lastWrittenTime,
+	}
+	cloned.size.Store(s.size.Load())
+	cloned.deltaRowcount.Store(s.deltaRowcount.Load())
+
+	for _, opt := range opts {
+		opt.UpdateSegment(cloned)
+	}
+	return cloned
+}
+
+func (s *SegmentInfo) IsDeltaLogExists(logID int64) bool {
+	for _, deltaLogs := range s.GetDeltalogs() {
+		for _, l := range deltaLogs.GetBinlogs() {
+			if l.GetLogID() == logID {
+				return true
+			}
+		}
+	}
+	return false
+}
+
+func (s *SegmentInfo) IsStatsLogExists(logID int64) bool {
+	for _, statsLogs := range s.GetStatslogs() {
+		for _, l := range statsLogs.GetBinlogs() {
+			if l.GetLogID() == logID {
+				return true
+			}
+		}
+	}
+	return false
+}
+
+func (s *SegmentInfo) GetCurrentRows() int64 {
+	if s == nil {
+		return 0
+	}
+	return s.currRows
+}
+
+func (s *SegmentInfo) IsCompacting() bool {
+	if s == nil {
+		return false
+	}
+	return s.isCompacting
+}
+
+func (s *SegmentInfo) GetAllocations() []*Allocation {
+	return s.allocations
+}
+
+func (s *SegmentInfo) GetLastFlushTime() time.Time {
+	return s.lastFlushTime
+}
+
+func (s *SegmentInfo) GetLastWrittenTime() time.Time {
+	return s.lastWrittenTime
+}
+
+// GetSegmentSize use cached value when segment is immutable
+func (s *SegmentInfo) GetSegmentSize() int64 {
+	if s.size.Load() <= 0 || s.GetState() != commonpb.SegmentState_Flushed {
+		var size int64
+		for _, binlogs := range s.GetBinlogs() {
+			for _, l := range binlogs.GetBinlogs() {
+				size += l.GetMemorySize()
+			}
+		}
+
+		for _, deltaLogs := range s.GetDeltalogs() {
+			for _, l := range deltaLogs.GetBinlogs() {
+				size += l.GetMemorySize()
+			}
+		}
+
+		for _, statsLogs := range s.GetStatslogs() {
+			for _, l := range statsLogs.GetBinlogs() {
+				size += l.GetMemorySize()
+			}
+		}
+		if size > 0 {
+			s.size.Store(size)
+		}
+	}
+	return s.size.Load()
+}
+
+// GetDeltaCount use cached value when segment is immutable
+func (s *SegmentInfo) GetDeltaCount() int64 {
+	if s.deltaRowcount.Load() < 0 || s.State != commonpb.SegmentState_Flushed {
+		var rc int64
+		for _, deltaLogs := range s.GetDeltalogs() {
+			for _, l := range deltaLogs.GetBinlogs() {
+				rc += l.GetEntriesNum()
+			}
+		}
+		s.deltaRowcount.Store(rc)
+	}
+	r := s.deltaRowcount.Load()
+	return r
+}
```

Comment:
deltaRowCount might be changed by L0 Compaction now, this cached value might be inaccurate.

Reply:
Now `Clone` shall invalidate cache